### ISSUE TITLE
Examples and comparisons?

### ISSUE THREAD

**Issue by alphapapa** [Alexander-Miller/pfuture#2]
Hi,

This package looks very neat, but I'm not sure exactly when I would use it instead of, e.g. async.el.  If you could give some examples and comparisons in the readme, it would be a big help.

Thanks.

**Comment by Alexander-Miller** [comment]
Pfuture is a) for shell commands and external processes and b) explicitly a future, not a callback. You use it when you will *eventually* need the output of some external process, and you are able to start the process early such that in executes in parallel to your elisp code. The best example is probably treemacs, which is what I wrote pfuture for. When you press TAB to expand a node in treemacs roughly the following things happen:

* A pfuture process for `git status --porcelain --ignored .` is started (it's really a python script that does a bit   more)
* In parallel to that process Emacs first queries the files and directories for the node you are expanding
* Then it filters them based on its settings
* then it segregates them into first directories then files
* then it sorts the two lists according the custom sort function
* then it creates the lists of strings that will eventually be inserted into the treemacs buffer
* for files that includes querying the right icons for their extensions
* only now does treemacs being waiting for the git process to complete so it can parse its output and highlight the   strings before rendering
* that's also the reason that treemacs has both ignored-file and pre-file-insert predicates - the latter is applied   once files' git status is available so you can opt not to display gitignored files

If you have that kind of use-case pfuture is just what you need. As a side note: I don't immediately insert my strings not just because I first need to highlight them. I've found that inserting many small strings is in fact much slower than concatenating them and inserting just one large string. That might be useful for your own packages.

As for async.el it's useful if you need callbacks or explicitly need to run elisp code and its additional overhead is negligeable for your use-case or there's something more complex going on than a simply starting a process and getting its results later.

I wrote a naive comparison and for me pfutue was quite a bit faster:

```
(-let [s 0]
 (garbage-collect)
 (setq s (float-time))
 (let* ((default-directory "/home/a/Documents/git/dotfiles")
        (future (pfuture-new "git" "status" "--porcelain")))
   (pfuture-await-to-finish future)
   (message "Pfuture done in %s" (- (float-time) s))
   (message "%s" (pfuture-result future))))

(progn
 (garbage-collect)
 (setq s (float-time))
 (async-start
  (lambda ()
    (let ((default-directory "/home/a/Documents/git/dotfiles"))
      (shell-command-to-string "git status --porcelain")))
  (lambda (res)
    (message "Async done in %s" (- (float-time) s))
    (message "%s" res))))

Pfuture done in 0.0040743350982666016
Async done in 0.1878223419189453
```

I'll try to add a short version of this to the readme.

**Comment by alphapapa** [comment]
Hi,

Thank you very much, that was very instructive!  I guess for my use case, pfuture isn't suitable, because I need a callback.  Although, is there any chance you could add a callback version?  That would be very useful.

I think you're right that pfuture is faster than async.el.  I ran your tests on my system:

```org
#+BEGIN_SRC elisp
  (list (list "pfuture" (-let [s 0]
                          (garbage-collect)
                          (setq s (float-time))
                          (let* ((default-directory "~/.emacs.d")
                                 (future (pfuture-new "git" "status" "--porcelain")))
                            (pfuture-await-to-finish future)
                            (pfuture-result future)
                            (- (float-time) s))))
        (list "async" (let ((future (async-start
                                     (lambda ()
                                       (let ((default-directory "~/.emacs.d"))
                                         (shell-command-to-string "git status --porcelain"))))))
                        (garbage-collect)
                        (setq s (float-time))
                        (async-wait future)
                        (- (float-time) s))))
#+END_SRC

#+RESULTS:
| pfuture |  0.01747727394104004 |
| async   |  0.05023717880249023 |
```

> As a side note: I don't immediately insert my strings not just because I first need to highlight them. I've found that inserting many small strings is in fact much slower than concatenating them and inserting just one large string. That might be useful for your own packages.

Now this puzzled me.  I'm not sure that's correct.  I ran some simple benchmarks and came up with this:

```org
#+BEGIN_SRC elisp
  (let ((strings (cl-loop for i from 1 to 1000
                          collect (number-to-string i))))
    (garbage-collect)
    (--sort (< (caddr it) (caddr other))
            (cl-loop for times in '(1 10 100)
                     append (a-list "Many small insertions"
                                    (cons times
                                          (benchmark-run-compiled times
                                            (with-temp-buffer
                                              (cl-loop for string in strings
                                                       do (insert string)))))
                                    "Concat and insert one string"
                                    (cons times
                                          (benchmark-run-compiled times
                                            (with-temp-buffer
                                              (cl-loop for string in strings
                                                       concat string into big-string
                                                       finally do (insert big-string)))))))))
#+END_SRC

#+RESULTS:
| Many small insertions        | 100 | 0.000708706 | 0 | 0.0 |
| Many small insertions        |  10 | 0.000714028 | 0 | 0.0 |
| Many small insertions        |   1 | 0.000924862 | 0 | 0.0 |
| Concat and insert one string |  10 | 0.003224029 | 0 | 0.0 |
| Concat and insert one string | 100 | 0.003256017 | 0 | 0.0 |
| Concat and insert one string |   1 | 0.003263461 | 0 | 0.0 |
```
It looks like small insertions are faster.  That makes sense to me, because concatting strings is extra work compared to just inserting them into buffers, which is relatively efficient.  Am I doing something wrong?  :)

Thanks also for the explanation about treemacs.  It sounds like I could learn a lot from its code.

**Comment by alphapapa** [comment]
Hi again,

I did indeed look at Treemacs' code, and I see that I was indeed doing something wrong.  This shows that you are right!

```org
#+BEGIN_SRC elisp
  (let ((strings (cl-loop for i from 1 to 1000
                          collect (number-to-string i))))
    (garbage-collect)
    (--sort (< (caddr it) (caddr other))
            (cl-loop for times in '(1 10 100)
                     append (a-list "(loop do (insert..."
                                    (cons times
                                          (benchmark-run-compiled times
                                            (with-temp-buffer
                                              (cl-loop for string in strings
                                                       do (insert string)))))
                                    "(apply #'insert..."
                                    (cons times
                                          (benchmark-run-compiled times
                                            (with-temp-buffer
                                              (apply #'insert strings))))
                                    "(insert (apply #'concat..."
                                    (cons times
                                          (benchmark-run-compiled times
                                            (with-temp-buffer
                                              (insert (apply #'concat strings)))))))))
#+END_SRC

#+RESULTS:
| (insert (apply #'concat... |  10 | 0.000107114 | 0 | 0.0 |
| (insert (apply #'concat... | 100 | 0.000109057 | 0 | 0.0 |
| (insert (apply #'concat... |   1 |  0.00010962 | 0 | 0.0 |
| (apply #'insert...         |  10 | 0.000607686 | 0 | 0.0 |
| (apply #'insert...         | 100 | 0.000608246 | 0 | 0.0 |
| (apply #'insert...         |   1 | 0.000627407 | 0 | 0.0 |
| (loop do (insert...        |  10 | 0.000664542 | 0 | 0.0 |
| (loop do (insert...        | 100 | 0.000674218 | 0 | 0.0 |
| (loop do (insert...        |   1 | 0.000926248 | 0 | 0.0 |
```
So thanks, I learned something there!  I feel like this should be documented as a best-practice somewhere...

**Comment by Alexander-Miller** [comment]
> Although, is there any chance you could add a callback version? That would be very useful.

Sure, it could act as a convenient wrapper around `make-process`. The way I envisage it, you'll be able to start a process that keeps collecting its output and then runs its callback with that output once its finished. Plus some decision about what to do in case of an error. Does that fit your use-case?

>  Am I doing something wrong?

I am not familiar with `cl-loop`, elisp-wise I grew up on dash. My guess would be that your code keeps concating every single string in the list one by one, allocating a new string every time, while the `apply` version calls concat only once. I suppose that many calls to `insert` accumulate enough overhead to make the concatenation worth it.

> So thanks, I learned something there!

Well here's another fun fact for you then: the speed of`directory-files` depends on the current major-mode, so its fastest when you wrap with `with-temp-buffer`.

As for documentation outside of style guides I am not aware of any such docs. I suppose I could at least put the info on /r/emacs.

>  It sounds like I could learn a lot from its code.

Oh I like talking about it ;)

**Comment by alphapapa** [comment]
> Sure, it could act as a convenient wrapper around make-process. The way I envisage it, you'll be able to start a process that keeps collecting its output and then runs its callback with that output once its finished. Plus some decision about what to do in case of an error. Does that fit your use-case?

Yes, I think that would be perfect for the code I'm working on here, which calls `ag` and `rg`: https://github.com/alphapapa/magit-todos/tree/async-ag

> I am not familiar with cl-loop, elisp-wise I grew up on dash. My guess would be that your code keeps concating every single string in the list one by one, allocating a new string every time, while the apply version calls concat only once.

I think that's right.  This code:

```el
(cl-loop for string in strings
         do (insert string))
```
expands to:
```el
(cl-block nil
  (let* ((--cl-var-- strings)
         (string nil)
         (big-string ""))
    (while (consp --cl-var--)
      (setq string (car --cl-var--))
      (cl-callf concat big-string string)
      (setq --cl-var-- (cdr --cl-var--)))
    (insert big-string)
    nil))
```
> I suppose that many calls to insert accumulate enough overhead to make the concatenation worth it.

I guess that's right too.  I thought that calling `insert` many times in a temp buffer, without any markers or overlays, wouldn't have significant overhead, but I guess if you call it 1000 times in a loop, it adds up.

> Well here's another fun fact for you then: the speed of directory-files depends on the current major-mode, so its fastest when you wrap with with-temp-buffer.

Now that is interesting.  I wonder why that is.  And I wonder if `f-files` has that optimization...

> As for documentation outside of style guides I am not aware of any such docs. I suppose I could at least put the info on /r/emacs.

I think I will add a "best practices" section in the "buffers" section of https://github.com/alphapapa/emacs-package-dev-handbook.  Feel free to suggest other things that should go there, or send PRs.  :)

**Comment by Alexander-Miller** [comment]
What do you think about this?

```
(cl-defmacro pfuture-callback (&key command on-complete on-error default-directory)
  `(let* ((default-directory ,default-directory)
          (process
          (make-process
           :name (concat "Pfuture Callback: " (string-join ,command " "))
           :command ,command
           :connection-type 'pipe
           :filter #'pfuture--append-output
           :sentinel (lambda (process status)
                       (let ((result (process-get process 'result)))
                         (message "Status change to [%s]/[%s]" status (process-exit-status process))
                         ,on-complete)))))
     process))
```

It still needs cleanup, error handling, docs and an edebug spec, but for the most part it should be useable. Example use would look like this:
```
(pfuture-callback
 :default-directory "/home/a/Documents/git/pfuture"
 :command '("git" "status")
 :on-complete (message "Result = %s" result))
```

The status in the sentinel can take values like `exited abnormally with code 1`, so determining whether a process completed successfully should be done with `process-exit-status`. I am also not a fan of how quoted lists look, so I'll also accept vectors for the `:command`, purely on aesthetic grounds.

**Comment by alphapapa** [comment]
That looks pretty good.  A few thoughts:

* `default-directory` is explicit, but since one of the ideas is convenience, shortening the keyword argument to `:directory` would be nice.  The docstring could explain that it actually binds `default-directory` around the call.
* It's probably good to default the connection-type to `pipe`, but it may sometimes be necessary to change it, so it would be good to have it as an argument.
* What is `pfuture--append-output`?  I can't find it anywhere.  :)
* This may just be my ignorance, but what exactly is `result` bound to?  i.e. is it a string, buffer, etc?  It would be nice to have it as an option, so e.g. getting a string could be the default, but if the user were to specify an argument like `:parser nil` (or whatever it would be called), the callback would be passed the buffer rather than a string (and then the callback would be responsible for killing the buffer).  That way another parser could also be specified, like a JSON-parsing function, that could return a parsed JSON object from the buffer (in which case `pfuture` should kill the buffer itself).  I realize this starts to sound complicated, but solving these problems in `pfuture` makes it painless for callers.  This is what I'm trying to do with `url-with-retrieve-async`, which solves some similar problems.
* I guess `on-error` is called when the process returns non-zero, but it's not clear whether `on-complete` would be called on a `0` exit code, or on mere "completion," which could be interpreted to mean "on both success and failure, whenever the process completes or exits."  If it's intended to only run on success, calling it `success` would be less ambiguous.
* The `:command '("git" "status")` argument is pretty nice, however it might be worth considering putting the command name in a separate argument, and using this keyword arg to specify arguments, e.g.:
```el
(pfuture-callback "git"
 :directory "/home/a/Documents/git/pfuture"
 :args '("status")
 :success (message "Result = %s" result))
```
That makes the most important argument, the command name, easy to see at a glance.  It could also be flexible enough to accept a list as the positional arg, so it could be called like:
```el
(pfuture-callback ("git" "status")
 :directory "/home/a/Documents/git/pfuture"
 :success (message "Result = %s" result))
```
Which makes it even easier to use for simple commands.

**Comment by Alexander-Miller** [comment]
> shortening the keyword argument to :directory would be nice.

Sure.

>  but it may sometimes be necessary to change it, so it would be good to have it as an argument

Optional arguments are also an option - no pun intended.

> What is pfuture--append-output?

Looks like it's an old change I never pushed. It's just [this lambda](https://github.com/Alexander-Miller/pfuture/blob/master/pfuture.el#L43) extracted.

> This may just be my ignorance, but what exactly is result bound to?

Right now result is the output of the process, saved as a string. There's no buffer, as I understand things. I remember trying to put the output into a buffer, thinking it should be faster since the output does not need to be concatenated multiple times, but iirc the results didn't really show an improvement. I'll try running some more tests, see if that still holds.

> calling it success would be less ambiguous.

Sounds good.

> It could also be flexible enough to accept a list as the positional arg,

Good idea.

**Comment by alphapapa** [comment]
Sounds great, I look forward to using this!

**Comment by Alexander-Miller** [comment]
Next version. Everything's in, aside from a buffer option. I added an `:on-status-change` option that'll run on every status change, even if the process hasn't finished, for debugging purposes. And also a `:name` option because why not. Thanks to macro-magic you don't need to quote the command, and a vector is ok, too.

Code:
```
(cl-defmacro pfuture-callback (command &key directory on-success on-error on-status-change name connection-type)
  (declare (indent 1))
  (let* ((command (if (vectorp command)
                      (cl-map 'list #'identity command)
                    command))
         (name (or name (concat "Pfuture Callback: [" (mapconcat #'identity command " ") "]")))
         (connection-type (or connection-type (quote 'pipe)))
         (directory (or directory default-directory)))
    `(let* ((default-directory ,directory)
            (process
             (make-process
              :name ,name
              :command ',command
              :connection-type ,connection-type
              :filter #'pfuture--append-output
              :sentinel (lambda (process status)
                          ,on-status-change
                          (unless (process-live-p process)
                            (let ((result (process-get process 'result)))
                              (if (= 0 (process-exit-status process))
                                  ,on-success
                                ,on-error)))))))
       process)))
```
And example use:
```
(pfuture-callback ["git" "status" "--porcelain"]
  :directory "~/Documents/git/dotfiles"
  :connection-type 'pipe
  :name "Test Pfuture"
  :on-status-change (message "Status Change to [%s]" (s-trim status))
  :on-success (message "Finished with output:\n%s" (s-trim result))
  :on-error (message "OH NOES!\n%s" (s-trim result)))
```

Hope that covers all your bases, I'm pretty sure it covers mine. There's just one small limitation: using a `cl-defmacro` means edebug is out, if you need it you'll have to call into your own named functions.

I've also pushed some small commits I had lying around so you can find `pfuture--append-output` on master.

**Comment by alphapapa** [comment]
Thanks, that looks really nice!  I'll dig into it some more soon when I have more time.

Curious about edebug though.  Can you not fix that with `(declare (debug ...))`?

**Comment by Alexander-Miller** [comment]
I thought edebug wouldn't work since the cl-defmacro declaration's argument list expands into `(&rest --cl-rest--)`, but it looks like I've been doing it wrong. It should work after all, I just need to wrap my head around edebug with keywork args.

**Comment by alphapapa** [comment]
> I just need to wrap my head around edebug with keywork args.

Good luck with that, whenever I need to do something with it that's not very simple, it's always a lot of trial and error.  I tried yesterday and finally gave up.  I think having more examples in the manual would help a lot.

The one tip I can give, which you probably already know, is that the keywords should be strings in the debug form, e.g.: `(declare (debug (... ":keyword-name" keyword-value)))`.  At least, that's one way to do it; there are probably others.  And maybe that isn't even appropriate, because it seems like that would require fixed argument order.

Sometime I'll grep through the Emacs source and see if I can find some good examples.  If you find any that I could add [here](https://github.com/alphapapa/emacs-package-dev-handbook), let me know.

**Comment by Alexander-Miller** [comment]
I've pushed an update now. There's extensive eldoc, but you already know how it works.

Couldn't get edebug to work after all, I guess it's due to the whole thing being asynchronous.

**Comment by alphapapa** [comment]
Thanks, sorry I forgot about this for a while.  I'll take a look at it again.

BTW, I posted the `directory-files` temp-buffer thing to https://github.com/rejeep/f.el/issues/93, hopefully it will be added there.  Do you think it should also be mentioned on emacs-devel?  Seems like a bug in `directory-files` as well.

**Comment by Alexander-Miller** [comment]
> Thanks, sorry I forgot about this for a while.

As have I. Lots of shiny new things over at treemacs.

> Do you think it should also be mentioned on emacs-devel?

I talked to Eli about it on reddit. He didn't really know what's going on either and recommended I benchmark things using perf. I haven't done that yet, but whenever I get around to it I'll post that to emacs devel.

**Comment by alphapapa** [comment]
Ok, I did a little experimenting trying to use `pfuture-callback` in `magit-todos`.  I ran into a few problems:

1.  It's a macro.  I'm already using a macro to build the call to it, so this complicates handling of the `command` argument at expansion time.  I made a function version of the macro, which avoids those problems.
2.  Then I ran into problems getting the process output.  I need access to the process's output buffer, but `pfuture-callback` passes the output as a string, and the process's buffer then appears to be empty.  So I tried inserting the output string into a temp buffer, but somehow that breaks newline handling, and my code that processes lines of output sees no matches.

Thanks.

**Comment by Alexander-Miller** [comment]
> his complicates handling of the command argument at expansion time

Why? The command argument is just a list of strings exactly as `make-process` expects.

> Then I ran into problems getting the process output.

I wanted to change that part anyway. The callback will now be invoked with the process buffer instead of its output, the buffer is bound under the name `pfuture-buffer`. The process buffer is also  automatically killed after `on-success` or `on-error` have finished running unless you provide your own buffer with the `:buffer` argument.

**Comment by Alexander-Miller** [comment]
*Finally* getting around to working through these.

I've now added some (non-exhaustive) examples for both entry points and mentioned async.el. I think that fulfills the original purpose of this ticket.
